### ISSUE TITLE
Play Media

### ISSUE THREAD

**Issue by pbuergi** [Strixx76/samsungwam#3]
Thank you very much for this integration. Works for me. Only Media ia m missing. Would be really nice if that would be implemented as welll ;)

**Comment by Strixx76** [comment]
I totally agree with you, and it is the next feature to be implemented after I implemented a function for keeping the connection alive. But I have very limited time to work on this.

**Comment by Jeremwhy** [comment]
Highly waited feature, thanks for your work !

**Comment by square-spade** [comment]
Is the keep alive a general function or specifically used for play media? I'm happy to start/help with the play media function (and grouping would be epic). 
Let me know if I can lighten your load at all :)

**Comment by Strixx76** [comment]
Update: Play media is now implemented in the underlying library and in the integration, but I need some more weeks for testing before releasing it. So far I have been able to test it with the radio stations integration and with local media and that seems to be working fine.
Before the end of the month I hope to be ready to release it.

**Comment by oliver10110** [comment]
Using v0.1.0rc1 nothing works, besides volume control. When trying to play anything, Home Assistant shows error "Failed to perform the action media_player/play_media. async_check_connection.<locals>.outer_wrapper() got an unexpected keyword argument 'extra'"

**Comment by Strixx76** [comment]
I found the bug. I had forgotten one commit when updating the "keep alive feature". Unfortunately I have just discovered another bug when playing DLNA or playing from your phone in the native app, which requires some changes in the underlying pywam package. I will try fix that as well before releasing next candidate.

**Comment by Strixx76** [comment]
v0.1.0rc2 is now available where the bug is fixed.

**Comment by Strixx76** [comment]
> I guess this version is not available in HACS, or? Is there any recommended procedure to install/test rc versions?

If you have added it as a custom repository it is available if you choose show beta versions. If it is already installed you have to go to HACS -> Integrations -> Samsung Wireless Audio. And the click on the three dots in the top right corner and choose Download again. Then you will be able to choose to show beta versions.

**Comment by llucax** [comment]
I just added it via the Download button in HACS, but listing all HACS entities I noted there was a disabled switch entity called "Pre-release" (`switch.samsung_wireless_audio_pre_release`). I enabled it, flipped the switch and voila! I could update the the rc.

After the update I could play media from TuneIn and some local DLNA server. Amazing! Thanks a lot!

**Comment by Jeremwhy** [comment]
Hey, i've installed v0.1.0rc2 but can't have the play media working. When switching on the wifi source and trying to play a media i don't have any error message but nothing happens on my HW-H750. Any idea ? Source/volumes controls works well

**Comment by Strixx76** [comment]
What media are you trying to play? What log level do you have? You should get a warning in the log if the stream you are trying to play is not supported and therefore not sent to the speaker. I have only tried it with [Radio Browser](https://www.home-assistant.io/integrations/radio_browser) and that works.

But I just tried to play a media file from local media in Home Assistant and I noticed that it will be detected as not supported even though it is. The problem is that the mime type in the header _('text/plain')_ is not correct. But I guess that is a Home Assistant bug!

I need to have some type of checks so that a none supported stream is sent to the speaker. The speaker has no checks what so ever, and will try to play whatever link you send it. And if it is not supported it has happened that the speaker freezes, and you need to unplug it to get it responding again. But i will see if I can fix some kind of workaround to support local media from Home Assistant the next week. I have some ideas.

But then there also is the risk that your speaker doesn't support that API. According that what I have found on the net, some speakers is not compatible, but I don't know how to tell which is, and which isn't.

**Comment by Strixx76** [comment]
There should now be a new RC available where the mime types reported by Home Assistant is trusted if the URL is pointing at your Home Assistant instance. But if the URL is pointing else where, either the file extension or the http is checked instead.

I will update README.md before doing a final release but just for info the API used to play URLs is a bit strange.

- The speaker is not reporting that a stream is being played. Which means that if you send a stream to the speaker and then starts the Multiroom app, your player will be in wrong state in both the app and HA, but the stream will still be playing.
- There seems also not to be a proper way to stop the stream after it is started. I have found one API working for pausing the stream, but resume doesn't work. And it has happened when a stream is paused that it starts by it self again after a couple of hours.
- Finally, since speaker doesn't report any info about the stream or state when playing a URL, there is now way for the integration to know when the stream has stopped.

Any ideas or input about how to fix the above is welcomed!

**Comment by pbuergi** [comment]
Thank you for your update.
DLNA is working, but only if it is in the same subnet (network). If it is in a different network (subnet) it does not play, even if the firewall is configured and the speaker can be accessed.

Play/Pause would obviously be nice if it works.

**Comment by Strixx76** [comment]
> Thank you for your update. DLNA is working, but only if it is in the same subnet (network). If it is in a different network (subnet) it does not play, even if the firewall is configured and the speaker can be accessed.

This has nothing to do with the integration. DLNA player is a function in the speaker.

**Comment by MrMEEE** [comment]
Hi @Strixx76 

Does this mean that I should be able to see my speakers from Music Assistanr (I can't)

**Comment by Strixx76** [comment]
@MrMEEE Yes. Your speakers will work in Music Assistant, but will not be discovered. You will have to add them you self:

1. Install Music Assistant Server (https://www.music-assistant.io/installation/)
2. Add the Music Assistant integration in Home Assistant (https://www.music-assistant.io/integration/)
3. Add Home Assistant as Player provider (https://www.music-assistant.io/player-support/ha/)
4. Go to settings for the Player provider and add you speakers

**Comment by MrMEEE** [comment]
> [@MrMEEE](https://github.com/MrMEEE) Yes. Your speakers will work in Music Assistant, but will not be discovered. You will have to add them you self:
> 
>     1. Install Music Assistant Server (https://www.music-assistant.io/installation/)
> 
>     2. Add the Music Assistant integration in Home Assistant (https://www.music-assistant.io/integration/)
> 
>     3. Add Home Assistant as Player provider (https://www.music-assistant.io/player-support/ha/)
> 
>     4. Go to settings for the Player provider and add you speakers
> 
> 
> Speakers that are grouped as slave will not be possible to add. So make sure all speakers are ungrouped when setting it up.

Thanks, that made them appear :)

**Comment by Strixx76** [comment]
Added in [v0.1.0](https://github.com/Strixx76/samsungwam/releases/tag/v0.1.0)
